tests/teststun.nim

```diff
@@ -0,0 +1,87 @@
+# Nim-WebRTC
+# Copyright (c) 2024 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+{.used.}
+
+import options, strutils
+import bearssl
+import ../webrtc/udp_connection
+import ../webrtc/stun/stun_connection
+import ../webrtc/stun/stun_message
+import ../webrtc/stun/stun_attributes
+import ../webrtc/stun/stun_utils
+import ./asyncunit
+
+proc newRng(): ref HmacDrbgContext =
+  HmacDrbgContext.new()
+
+suite "Stun message encoding/decoding":
+  test "Get BindingRequest + encode & decode with a set username":
+    let
+      udpConn = UdpConn.init(AnyAddress)
+      conn = StunConn.init(udpConn, TransportAddress(AnyAddress), true, newRng())
+      msg = conn.getBindingRequest(username = "Test")
+      encoded = msg.encode(msg.getAttribute(AttrUsername))
+      decoded = StunMessage.decode(encoded)
+
+    check:
+      msg.msgType == decoded.msgType
+      msg.transactionId == decoded.transactionId
+      msg.getAttribute(AttrUsername) == decoded.getAttribute(AttrUsername)
+      msg.getAttribute(AttrICEControlling).isSome()
+      msg.getAttribute(AttrICEControlled).isNone()
+      Priority.decode(msg.getAttribute(AttrPriority).get()).priority == 0x7effffff
+      # encoding adds Fingerprint as attributes
+      msg.attributes.len() == decoded.attributes.len() - 1
+    conn.close()
+
+  test "Get BindingRequest + encode & decode with a random username":
+    let
+      udpConn = UdpConn.init(AnyAddress)
+      conn = StunConn.init(udpConn, TransportAddress(AnyAddress), false, newRng())
+      msg = conn.getBindingRequest()
+      encoded = msg.encode(msg.getAttribute(AttrUsername))
```

Comment:
Could you get the username attribute from the msg inside the encode proc?

Reply:
Yes, you can. But I implemented `encode` with libp2p in mind.
In theory, you don't need `Username` to encode. You need a password. But with libp2p, the username looks like `libp2p+webrtc+v1/<random string>` and the password we use is the random string.
So if I implemented `encode` this way, is because I think of a possible future where we can use any password instead of using `Username`.

Reply:
I'm not sure if I get it. What does the spec say exactly? Also, we definitely shouldn't have a string like `"libp2p+webrtc+v1/"` in this code as it is independent of libp2p.

Reply:
shouldn't libp2p create the request with the right value in the attribute?

Reply:
I see we already have the option of specifying a username in `getBindingRequest` (although I'm still confused with this username and password thing. We should remove this `"libp2p+webrtc+v1/"` part. Most likely `username` is a required field and this logic we are currently doing when it isn't specified must be done by the caller.

Reply:
> What does the spec say exactly?

It's basically a mix between the implementation of Message-Integrity of the RFC https://datatracker.ietf.org/doc/html/rfc5389#section-15.4 and the libp2p spec: https://github.com/libp2p/specs/blob/master/webrtc/webrtc-direct.md?plain=1#L73-L77.
> Also, we definitely shouldn't have a string like "libp2p+webrtc+v1/" in this code as it is independent of libp2p.

I agree. I have to think about it though. The problem is the username should be a new one for every connection. So maybe two callbacks `getUsername()` and `getPassword()` could do the trick.
> shouldn't libp2p create the request with the right value in the attribute?

At the moment, libp2p only interact with DataChannel. But with the two callbacks, it should be possible.
> I see we already have the option of specifying a username in getBindingRequest (although I'm still confused with this username and password thing. We should remove this "libp2p+webrtc+v1/" part. Most likely username is a required field and this logic we are currently doing when it isn't specified must be done by the caller.

I agree. The more I think about it the more the two callbacks make sense. And that way, it's completely independent from libp2p.

Reply:
I totally reworked this part with three callback functions to get username / check username / get password.